### PR TITLE
Feat/MET-5193 Script for Performance Metrics

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    private static final String OUTPUT_FILE_NAME = "performance-metric-";
+
+    public static void main(String[] args) throws CustomTruststoreAppender.TrustStoreConfigurationException, FileNotFoundException, ParseException {
+        LOGGER.info("Starting script");
+
+        final MongoMetisCoreDao mongoMetisCoreDao = new MongoMetisCoreDao(propertiesHolder);
+        final CSVUtilities csvUtilities = new CSVUtilities(mongoMetisCoreDao);
+
+        //Prepare input values
+        String startDateAsString = "2022-02-15 00:00:00";
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java]: this ones I see they remain constant we can add the final modifier
```
        final String startDateAsString = "2022-02-15 00:00:00";
        final String endDateAsString = "2023-04-01 23:59:59";
        final SimpleDateFormat simpleDateFormatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        final Date startDate = simpleDateFormatter.parse(startDateAsString);
        final Date endDate = simpleDateFormatter.parse(endDateAsString);
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.FileNotFoundException;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+
+public class PerformanceMetricMain {
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java:20]: Include java doc.

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+        //Prepare input values
+        String startDateAsString = "2022-02-15 00:00:00";
+        String endDateAsString = "2023-04-01 23:59:59";
+        SimpleDateFormat simpleDateFormatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+        Date startDate = simpleDateFormatter.parse(startDateAsString);
+        Date endDate = simpleDateFormatter.parse(endDateAsString);
+
+        //Start printing
+        csvUtilities.writeIntoCsvFile(FILE_PATH + OUTPUT_FILE_NAME + NOW_DATE.format(ISO_FORMATTER) + ".csv", startDate, endDate);
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java:45]: Suggestion to  add a validation, if the path doesn't exist that it creates the path if it doesn't exist or handles the exception.

```
Exception in thread "main" java.io.FileNotFoundException: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/output/performance-metric-2023-04-24T14:04:21.072376.csv (Bestand of map bestaat niet)
	at java.base/java.io.FileOutputStream.open0(Native Method)
	at java.base/java.io.FileOutputStream.open(FileOutputStream.java:298)
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:237)
	at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:187)
	at java.base/java.io.PrintWriter.<init>(PrintWriter.java:309)
	at eu.europeana.metis.performance.metric.utilities.CSVUtilities.writeIntoCsvFile(CSVUtilities.java:41)
	at eu.europeana.metis.performance.metric.PerformanceMetricMain.main(PerformanceMetricMain.java:42)
```

File: metis-performance-metric/src/main/resources/log4j2.xml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="WARN">
+  <Appenders>
+    <Console name="Console" target="SYSTEM_OUT">
+      <PatternLayout pattern="%d{yyyy-MM-dd'T'HH:mm:ss.SSSXXX} %-5p %C:%L [%t] - %m%n"/>
+    </Console>
+    <File name="ExecutionLogs"
+      fileName="logs/metis-reprocessing/execution-${date:yyyy-MM-dd-HHmmss}.log">
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/resources/log4j2.xml]: suggestion: we should use a folder for metrics logs
```
logs/metis-metrics/execution-
logs/metis-metrics/statistics-
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CSVUtilities.class);
+
+    private final MongoMetisCoreDao mongoMetisCoreDao;
+    public CSVUtilities(MongoMetisCoreDao mongoMetisCoreDao) {
+        this.mongoMetisCoreDao = mongoMetisCoreDao;
+    }
+
+    public void writeIntoCsvFile(String filePath, Date startDate, Date endDate) throws FileNotFoundException {
+        File resultFile = new File(filePath);
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java]:
```
        final File resultFile = new File(filePath);
        final String firstRow = "DatasetId, Date Publish Indexing, Time of last successful harvest in hours, Number of Records Published";
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoInitializer.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  private final int[] mongoPorts;
+  private final String mongoAuthenticationDb;
+  private final String mongoUsername;
+  private final String mongoPassword;
+  private final boolean mongoEnablessl;
+  private final int metisCoreConnectionPoolSize;
+  private MongoClient mongoClient;
+
+  public MongoInitializer(String[] mongoHosts, int[] mongoPorts, String mongoAuthenticationDb,
+                          String mongoUsername, String mongoPassword, boolean mongoEnablessl, int metisCoreConnectionPoolSize) {
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoInitializer.java]: suggestion `mongoEnableSSL`

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoMetisCoreDao.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        metisCoreMongoInitializer = prepareMetisCoreConfiguration();
+        final MorphiaDatastoreProviderImpl morphiaDatastoreProvider = new MorphiaDatastoreProviderImpl(
+                metisCoreMongoInitializer.getMongoClient(), propertiesHolder.metisCoreMongoDb);
+        workflowExecutionDao = new WorkflowExecutionDao(morphiaDatastoreProvider);
+        dataEvolutionUtils = new DataEvolutionUtils(workflowExecutionDao);
+    }
+
+    public WorkflowExecutionDao.ResultList<WorkflowExecutionDao.ExecutionDatasetPair> getAllWorkflowsExecutionsOverview(Date startDate, Date endDate){
+        //Make start date two weeks earlier, so we can include more reports based on end date
+        LocalDateTime startLocalDateTime = LocalDateTime.ofInstant(startDate.toInstant(), ZoneId.systemDefault());
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoMetisCoreDao.java]: this line can be rewrite including the final modifier 
```
final LocalDateTime startLocalDateTime = LocalDateTime
                .ofInstant(startDate.toInstant(), ZoneId.systemDefault())
                .minusWeeks(2L);
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+import java.io.PrintWriter;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class CSVUtilities {
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java:28]: include java doc for elements.

**Review by jeortizquan — COMMENTED** [review]
Have a look on the comments.

I tested with the following parameters.
01-01-2022 to 31-12-2022
Is it valid to have duplicate dataset Id's? 

```
DatasetId, Date Publish Indexing, Time of last successful harvest in hours, Number of Records Published
59, 2022-08-26 09:10:35, 0, 96
41, 2022-08-26 09:46:37, 0, 4
59, 2022-08-29 10:20:34, 0, 96
41, 2022-08-31 14:04:03, 0, 4
41, 2022-09-02 10:03:05, 0, 4
60, 2022-09-05 09:46:34, 0, 30
58, 2022-09-05 11:44:33, 0, 1
61, 2022-09-05 16:33:53, 0, 1
60, 2022-09-05 16:42:20, 0, 30
58, 2022-09-05 16:50:19, 3, 1
60, 2022-09-06 14:34:52, 0, 30
62, 2022-09-19 14:33:24, 0, 152
63, 2022-10-10 19:50:30, 4, 3398
64, 2022-10-10 19:53:30, 4, 152
64, 2022-10-11 16:28:30, 0, 152
64, 2022-10-18 15:48:00, 0, 152
64, 2022-10-20 17:11:00, 0, 152
65, 2022-10-21 14:41:46, 0, 21
58, 2022-10-24 16:04:42, 0, 3
58, 2022-10-24 16:15:45, 0, 3
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java (hunk 1 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java]: Fixed

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java (hunk 2 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java:20]: Done

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java (hunk 3 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java:45]: It should create a file automatically into `output` folder, did you get this exception?

File: metis-performance-metric/src/main/resources/log4j2.xml (hunk 4 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/resources/log4j2.xml]: You are right, let me fix it

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java (hunk 5 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java]: Fixed

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoInitializer.java (hunk 6 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoInitializer.java]: Done

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoMetisCoreDao.java (hunk 7 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/dao/MongoMetisCoreDao.java]: Fixed

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java (hunk 8 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java:28]: Done

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java (hunk 3 above)
Comment (jeortizquan) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java:45]: I hadn't the `output` folder that is why I got the exception :sweat_smile:

**Review by jeortizquan — COMMENTED** [review]
Code wise is working now. somehow is not showing all the published datasets. I have created this one in test env. If you see the results is not reflected in the report.

afbeelding

```
DatasetId, Date Publish Indexing, Time of last successful harvest in hours, Number of Records Published
156, 2023-04-04 11:21:47, 0, 1
157, 2023-04-04 11:18:45, 0, 2
160, 2023-04-04 11:16:57, 0, 5
163, 2023-04-04 16:04:04, 0, 0
165, 2023-04-05 13:47:13, 0, 1
166, 2023-04-05 13:58:11, 0, 1
167, 2023-04-05 13:54:54, 0, 1
168, 2023-04-05 16:32:25, 0, 1
169, 2023-04-06 09:07:13, 0, 1
170, 2023-04-11 15:13:42, 0, 1
171, 2023-04-11 17:19:44, 1, 1004
173, 2023-04-12 11:34:15, 1, 134
174, 2023-04-12 11:24:12, 0, 5
175, 2023-04-12 11:27:44, 0, 2
177, 2023-04-12 11:40:46, 0, 1
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java (hunk 3 above)
Comment (JoanaCMS) [reply metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/PerformanceMetricMain.java:45]: I noticed that the empty folder `output` was not pushed, so I added an example file inside it so git recognizes the folder, it shouldn't give problems now :)

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            LOGGER.error("Something went wrong when extracting data");
+            return "";
+        }
+
+        LOGGER.info("Finished processing dataset with id " + datasetId);
+
+        return stringBuilderCSVRow.toString();
+    }
+
+    private long calculateTimeDifference(WorkflowExecution workflowExecution, AbstractExecutablePlugin<?> publishPlugin){
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java:85]: One more suggestion to give a sense of time in hours using decimals, if a harvesting lasted minutes is displayed '0' zero
using a double will give for example 0.5 hrs
```
private double calculateTimeDifference(WorkflowExecution workflowExecution, AbstractExecutablePlugin<?> publishPlugin){
        PluginWithExecutionId<? extends ExecutablePlugin> harvestPlugin =
                mongoMetisCoreDao.getHarvesting(new PluginWithExecutionId<>(workflowExecution,publishPlugin));
        return TimeUnit.MILLISECONDS.toMinutes(publishPlugin.getFinishedDate().getTime() - harvestPlugin.getPlugin().getStartedDate().getTime()) / 60.0;
    }
```

File: metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        LOGGER.info("Processing dataset with id " + datasetId);
+        StringBuilder stringBuilderCSVRow = new StringBuilder();
+        stringBuilderCSVRow.append(datasetId);
+        stringBuilderCSVRow.append(", ");
+        Optional<AbstractMetisPlugin> optionalPublishPlugin = execution.getMetisPluginWithType(PluginType.PUBLISH);
+        if(optionalPublishPlugin.isPresent()){
+            AbstractMetisPlugin publishPlugin = optionalPublishPlugin.get();
+            stringBuilderCSVRow.append(simpleDateFormatter.format(publishPlugin.getFinishedDate()));
+            stringBuilderCSVRow.append(", ");
+            stringBuilderCSVRow.append(calculateTimeDifference(execution, (AbstractExecutablePlugin<?>)publishPlugin));
```
Comment (jeortizquan) [inline metis-performance-metric/src/main/java/eu/europeana/metis/performance/metric/utilities/CSVUtilities.java:72]: suggestion to format the result to two decimals.
```
stringBuilderCSVRow.append(String.format(Locale.ENGLISH,"%.2f",
                    calculateTimeDifference(execution, (AbstractExecutablePlugin<?>)publishPlugin)));
```

**Review by jeortizquan — COMMENTED** [review]
Have a look on this last suggestion ideas. so the outcome will look like this.
```
DatasetId, Date Publish Indexing, Time of last successful harvest in hours, Number of Records Published
2, 2022-05-17 15:54:42, 0.42, 3
4, 2023-04-25 15:32:08, 8231.22, 3
5, 2022-05-18 16:13:56, 0.08, 96
6, 2022-06-30 12:16:26, 0.03, 30
7, 2022-10-13 18:51:01, 0.05, 30
```
because without this rounding looks like this
```
DatasetId, Date Publish Indexing, Time of last successful harvest in hours, Number of Records Published
2, 2022-05-17 15:54:42, 0, 3
4, 2023-04-25 15:32:08, 8231, 3
5, 2022-05-18 16:13:56, 0, 96
6, 2022-06-30 12:16:26, 0, 30
7, 2022-10-13 18:51:01, 0, 30
```

**Review by jeortizquan — APPROVED** [review]
After discussing it will remain this version without decimals in the hours.

**Pull request merged**